File: src/main/scala/com/vc/medium/L1011.scala
package com.vc.medium

object L1011 {
  def shipWithinDays(weights: Array[Int], D: Int): Int = {
    var left = 0
    var right = 0
    weights.foreach(w => {
      left = Math.max(w, left)
      right += w
    })
    while(left < right) {
      val mid = left + (right - left) / 2
      var cur = 0
      var need = 1
      weights.foreach(w => {
        if(cur + w > mid) {
          need += 1
          cur = 0
        }
        cur += w
      })
      if(need > D) left += 1
      else right = mid
    }
    left
  }
}

File: src/main/scala/com/vc/medium/L1011.java
package com.vc.medium;

class L1011 {
    public int shipWithinDays(int[] arr, int D) {
        int lo = Integer.MIN_VALUE;
        int hi = 0;

        for(int i = 0; i < arr.length; i++) {
            lo = Math.max(arr[i], lo);
            hi += arr[i];
        }

        /**
            Minimize the maximum
         */
        while(lo < hi){
            int mid = lo + (hi - lo) / 2;
            boolean isPossible = isPossible(arr, mid, D);
            //System.out.println("Mid: "+mid+" isPossible: "+isPossible);
            if(!isPossible) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }

    private boolean isPossible(int[] arr, int weight, int D) {
        int total = 0;
        int count = 1;
        for(int i = 0; i < arr.length; i++) {
            total += arr[i];
            if(total > weight) {
                total = arr[i];
                count++;
                if(count > D) return false;
            }
        }
        return true;
    }
}